rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,177 @@
+7. The user's [configured Common Expression Language rule](#configuration) for the token will be evaluated against the claims, to ensure that the token is allowed to register with the Teleport cluster.
+8. Certificates will be generated for the client. The generated certificates will be non-renewable, as the client will proceed through the same steps to generate new certificates. This prevents exfiltratred credentials being used to repeatedly generate more credentials, maintaining access to the system.
+
+We will re-use `go-jose@v2` for validation of JWTs, since this library is already in use within Teleport.
+
+#### Caching JWKs
+
+Special attention should be given to the logic around caching the JWKs.
+
+We will cache these for two reasons:
+
+- Improve the performance of validating JWTs, as we will not need to make a HTTPS request to the issuer.
+- Improve the reliability, as we can validate JWTs even if the issuer is experiencing some downtime.
+- Reduce the impact of Teleport on an issuer. If onboarding a large number of nodes, we do not want to unduly place pressure on the issuer. 
+
+We should keep in mind the following considerations:
+
+- When we are presented with a JWT with a previously unseen `kid`, we should re-check the issuer's JWKs, as they may have begun issuing tokens with a new JWK.
+- We should ensure that the TTL of the cache entries is relatively short, as we want to allow an issuer to revoke a JWK that has been stolen.
+
+#### Configuration
+
+We will leverage the existing Token configuration kind as used by static tokens and IAM joining:
+
+```yaml
+kind: token
+version: v2
+metadata:
+  name: my-gce-token
+  expires: "3000-01-01T00:00:00Z"
+spec:
+  roles: [Node]
+  join_method: oidc-jwt
+  issuer_url: https://accounts.google.com
+  allow: claims.aud == "noah.teleport.sh" && claims.google.compute_engine.project_id == "my-project" && claims.google.compute_engine.instance_name == "an-instance"
+```
+
+To allow the user to configure rules for what identities will be accepted, we will use the [Common Expression Language (CEL)](https://github.com/google/cel-spec). This allows a large degree of flexibility in the complexity of rules users can configure, but still allows simple expressions.
+
+Users must also configure the `issuer_url`. This must be a host on which there is a compliant `/.well-known/openid-configuration` endpoint.
+
+### Node support
+
+Node here not only refers to a Teleport node, but also to various other participants within a Teleport cluster (e.g tbot, kube agent etc).
+
+We will need to support collecting the token from the environment. This will differ on each platform. Some offer the token via a metadata service, and others directly inject it via an environment variable. Where possible, we should encourage the user to configure the token to be generated with an audience of their Teleport cluster, however, not all providers support this (e.g GitLab CI/CD).
```

Comment:
I've certainly not seen any other providers doing this, but I can certainly understand the merit of it. The audience field would unfortunately be the only choice, since that's the only one that some providers let you specify when generating the token. However, I'm not sure that this brings enough merit to be worth adding, especially as specifying a custom audience is not supported on all platforms.